File: Java/src/_1049_Last_Stone_Weight_II/Solution.java
package _1049_Last_Stone_Weight_II;

/**
 * Date: 2021/6/8 9:50
 * 执行用时：2 ms, 在所有 Java 提交中击败了99.07%的用户
 * 内存消耗：35.3 MB, 在所有 Java 提交中击败了99.35%的用户
 */
public class Solution {

    public int lastStoneWeightII(int[] stones) {
        int sum = 0;
        for (int x : stones) sum += x;
        int m = sum / 2;
        int[] f = new int[m + 1];
        for (int x : stones)
            for (int j = m; j >= x; j--)
                f[j] = Math.max(f[j], f[j - x] + x);
        return (sum - f[m]) - f[m];
    }

    public static void main(String[] args) {

        int[] stones = {2, 7, 4, 1, 8, 1};
        System.out.println((new Solution()).lastStoneWeightII(stones));  // 1
    }
}

File: Python/_1049_Last_Stone_Weight_II/Solution.py
# coding=utf-8
# Date: 2021/6/8 9:52
from typing import List


# 执行用时：68 ms, 在所有 Python3 提交中击败了57.92%的用户
# 内存消耗：15 MB, 在所有 Python3 提交中击败了38.66%的用户
class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        s = sum(stones)
        m = s // 2
        f = [0 for _ in range(m + 1)]
        for x in stones:
            for j in range(m, x - 1, -1):
                f[j] = max(f[j], f[j - x] + x)
        return s - 2 * f[m]


if __name__ == "__main__":
    print(Solution().lastStoneWeightII([2, 7, 4, 1, 8, 1]))  # 1
